nifi-nar-bundles/nifi-social-media-bundle/nifi-twitter-processors/src/main/java/org/apache/nifi/processors/twitter/TweetStreamService.java

```diff
@@ -0,0 +1,222 @@
+        } else {
+            logger.warn("Unrecognized endpoint in getTransitUri. Returning basePath");
+            return api.getApiClient().getBasePath();
+        }
+    }
+
+    /**
+     * This method would be called when we would like the stream to get started. This method will spin off a thread that
+     * will continue to queue tweets on to the given queue passed in the constructor. The thread will continue
+     * to run until {@code stop} is called.
+     * @param endpoint {@code TwitterStreamAPI.SAMPLE_ENDPOINT} or {@code TwitterStreamAPI.SEARCH_ENDPOINT}
+     */
+    public void start(final StreamEndpoint endpoint) {
+        this.endpoint = endpoint;
+        executorService.execute(new TweetStreamStarter());
+    }
+
+    /**
+     * This method would be called when we would like the stream to get stopped. The stream will be closed and the
+     * executorService will be shut down.
+     */
+    public void stop() {
+        if (stream != null) {
+            try {
+                stream.close();
+            } catch (IOException e) {
+                logger.error("Closing response stream failed", e);
+            }
+        }
+
+        executorService.shutdownNow();
+    }
+
+    private Long calculateBackoffDelay() {
+        long backoff = backoffMultiplier * backoffTime;
+        long delay = (backoff > maximumBackoff) ? maximumBackoff : backoff;
+        return delay;
+    }
+
+    private void scheduleStartStreamWithBackoff() {
+        // use exponential(by factor of 2) backoff in scheduling the next TweetStreamStarter
+        if (attemptCounter >= backoffAttempts) {
+            throw new ProcessException(String.format("Connection failed after maximum attempts [%d]", attemptCounter));
+        }
+        attemptCounter += 1;
+        long delay = calculateBackoffDelay();
+        backoffMultiplier *= 2;
+        logger.info("Scheduling new stream connection after delay [{} s]", delay);
+        executorService.schedule(new TweetStreamStarter(), delay, TimeUnit.SECONDS);
+    }
+
+    private void resetBackoff() {
+        attemptCounter = 0;
+        backoffMultiplier = 1L;
+    }
+
+    private class TweetStreamStarter implements Runnable {
+        @Override
+        public void run() {
+            try {
+                if (endpoint.equals(StreamEndpoint.SAMPLE_ENDPOINT)) {
+                    stream = api.tweets().sampleStream(expansions, tweetFields, userFields, mediaFields, placeFields, pollFields, backfillMinutes);
+                } else {
+                    stream = api.tweets().searchStream(expansions, tweetFields, userFields, mediaFields, placeFields, pollFields, backfillMinutes);
+                }
+                executorService.execute(new TweetStreamHandler());
+            } catch (ApiException e) {
+                stream = null;
+                logger.warn("Twitter Stream [{}] API connection failed: HTTP {}", endpoint.getEndpointName(), e.getCode(), e);
```

Comment:
Unfortunately the `ApiException` does not include the response body as part of the standard message, so adjusting the log to include it would be helpful:
```suggestion
                logger.warn("Twitter Stream [{}] API connection failed: HTTP {} {}", endpoint.getEndpointName(), e.getCode(), e.getResponseBody(), e);
```